File: api/js/src/validator/rules/incompatible-button-context.ts
import { createDiagnostic } from "../diagnostics.js";
import type { ValidationRule } from "../types.js";
import { RULE_IDS } from "./constants.js";
import { buttonKind, loc } from "./helpers.js";

export const incompatibleButtonContextRule: ValidationRule = {
  id: RULE_IDS.INCOMPATIBLE_BUTTON_CONTEXT,
  description: "Buttons must match the declared validation context",
  defaultSeverity: "error",
  run(ctx) {
    const diagnostics = [];
    const contextType = ctx.contextType;

    for (const ref of ctx.normalized.flat) {
      const { button, rowIndex, columnIndex, flatIndex } = ref;
      const kind = buttonKind(button);

      if (kind === "pay" && contextType !== "invoice") {
        diagnostics.push(
          createDiagnostic(
            RULE_IDS.INCOMPATIBLE_BUTTON_CONTEXT,
            "Pay buttons are only valid in invoice keyboards",
            "error",
            loc(rowIndex, columnIndex, flatIndex),
            'Set validate({ contextType: "invoice" }) when building payment keyboards',
          ),
        );
      }

      const hasCallback =
        "callback_data" in button && button.callback_data !== undefined;
      const hasUrl = "url" in button && button.url !== undefined;
      const hasPay = "pay" in button && button.pay === true;

      const actionCount = [hasCallback, hasUrl, hasPay].filter(Boolean).length;
      if (actionCount > 1) {
        diagnostics.push(
          createDiagnostic(
            RULE_IDS.INCOMPATIBLE_BUTTON_CONTEXT,
            "Button defines multiple mutually exclusive actions",
            "error",
            loc(rowIndex, columnIndex, flatIndex),
          ),
        );
      }

      if (kind === "custom" && contextType === "invoice") {
        diagnostics.push(
          createDiagnostic(
            RULE_IDS.INCOMPATIBLE_BUTTON_CONTEXT,
            "Custom buttons without url/callback/pay are unusual in invoice context",
            "warning",
            loc(rowIndex, columnIndex, flatIndex),
          ),
        );
      }
    }

    return diagnostics;
  },
};

File: api/python/telegram_inline_keyboard_builder/validator/rules/incompatible_button_context.py
from __future__ import annotations

from ..diagnostics import create_diagnostic
from ..types import Diagnostic, RuleContext, ValidationRule
from .constants import RULE_IDS
from .helpers import button_kind, loc


def _run(ctx: RuleContext) -> list[Diagnostic]:
    diagnostics: list[Diagnostic] = []
    context_type = ctx.context_type

    for ref in ctx.normalized.flat:
        button = ref.button
        kind = button_kind(button)

        if kind == "pay" and context_type != "invoice":
            diagnostics.append(
                create_diagnostic(
                    RULE_IDS.INCOMPATIBLE_BUTTON_CONTEXT,
                    "Pay buttons are only valid in invoice keyboards",
                    "error",
                    loc(ref.row_index, ref.column_index, ref.flat_index),
                    'Set validate(context_type="invoice") when building payment keyboards',
                )
            )

        has_callback = (
            "callback_data" in button and button.get("callback_data") is not None
        )
        has_url = "url" in button and button.get("url") is not None
        has_pay = button.get("pay") is True
        action_count = sum([has_callback, has_url, has_pay])
        if action_count > 1:
            diagnostics.append(
                create_diagnostic(
                    RULE_IDS.INCOMPATIBLE_BUTTON_CONTEXT,
                    "Button defines multiple mutually exclusive actions",
                    "error",
                    loc(ref.row_index, ref.column_index, ref.flat_index),
                )
            )

        if kind == "custom" and context_type == "invoice":
            diagnostics.append(
                create_diagnostic(
                    RULE_IDS.INCOMPATIBLE_BUTTON_CONTEXT,
                    "Custom buttons without url/callback/pay are unusual in invoice context",
                    "warning",
                    loc(ref.row_index, ref.column_index, ref.flat_index),
                )
            )

    return diagnostics


incompatible_button_context_rule = ValidationRule(
    id=RULE_IDS.INCOMPATIBLE_BUTTON_CONTEXT,
    description="Buttons must match the declared validation context",
    default_severity="error",
    run=_run,
)
